### PR TITLE
Obtain MessageDigest via FastThreadLocal

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Creating a new MessageDigest every time is wasteful, we should store them in FastThreadLocal.

Modifications:

Change WebSocketUtil to store MD5 and SHA1 MessageDigest in FastThreadLocal and use these.

Result:

Less overhead and less GC.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @trustin PTAL

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketUtil.java [hunk 1]
```diff
@@ -52,15 +72,11 @@
...
-            //Hash the data
-            return md.digest(data);
-        } catch (NoSuchAlgorithmException e) {
-            //Alright, you might have an old system.
-            throw new InternalError("SHA-1 is not supported on this platform - Outdated?");
-        }
+        return hash(SHA1, data);
+    }
+
+    private static byte[] hash(FastThreadLocal<MessageDigest> mds, byte[] data) {
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketUtil.java]: consider just inlining this.

Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketUtil.java]: sure why not...

**Comment by trustin** [comment]
How about just writing a utility class dedicated to producing md5/sha1 hash? Given that what you really want is a single-array-in-single-array-out digest, the utility class can be completely stateless and thread-safe without any lock. i.e. no reason to use thread local.

**Comment by normanmaurer** [comment]
Sure... Let me give this a try once I have a bit time.

**Comment by normanmaurer** [comment]
@trustin can we do this as a follow up pr as this will take some more time and this PR already gives us an improvement. WDYT ?

**Comment by trustin** [comment]
@normanmaurer Could you please add `TODO(normanmaurer)` then?

**Comment by normanmaurer** [comment]
Will do and open an issue

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (f8f58cd73449c27ecf4637072a0127f647392b24) and 4.1 (fa947643a0a025d29976f4d149fc0d298f0e4bed)

**Comment by normanmaurer** [comment]
See https://github.com/netty/netty/issues/4641

**Pull request closed without merging**